include/boost/wave/util/cpp_macromap.hpp

```diff
@@ -973,19 +982,29 @@ bool adjacent_stringize = false;
         typename ContainerT::size_type i;
 #if BOOST_WAVE_SUPPORT_VARIADICS_PLACEMARKERS != 0
         bool is_ellipsis = false;
+        bool is_va_opt = false;
 
             if (IS_EXTCATEGORY((*cit), ExtParameterTokenType)) {
                 BOOST_ASSERT(boost::wave::need_variadics(ctx.get_language()));
                 i = token_id(*cit) - T_EXTPARAMETERBASE;
                 is_ellipsis = true;
             }
             else
+#if BOOST_WAVE_SUPPORT_VA_OPT != 0
+
+            if (IS_EXTCATEGORY((*cit), OptParameterTokenType)) {
+                BOOST_ASSERT(boost::wave::need_va_opt(ctx.get_language()));
```

Comment:
I think this should actually be an error. Otherwise applications will silently fail if they configure the library for C++11 but encounter a `__VA_OPT__` (I have seen this assertion fire in the `wave --c++11 ...`, which is unexpected and obscure).

Reply:
I am following the pattern from [this line](https://github.com/jefftrull/wave/blob/60ebbd80e35cd8a7fffc71870f7be22e3a0ad4ce/include/boost/wave/util/cpp_macromap.hpp#L988); is `__VA_OPT__` a different situation in this way?

Reply:
Right, it's probably wrong for `__VA_ARGS__` above as well. :/

Reply:
I'll add unit tests and modify the checking code.

Reply:
I have an alternative proposal:
Let's keep the assert. "This should never happen" is actually a sensible idea here, because `T_EXTRAPARAMETER`/`T_OPTPARAMETER` are marked by a previous parsing step.
Instead, maybe we can add a check for the language [at that prior point](https://github.com/jefftrull/wave/blob/f6f71ddb8cde9b99dd36db79beb75629e4c779a1/include/boost/wave/util/macro_definition.hpp#L111) - a context is available for passing into `replace_parameters()`...

What do you think?

Reply:
Sounds good to me! Thanks!

Reply:
OK so I'm back again :)
I can reproduce this issue with `__VA_OPT__` but not `__VA_ARGS__`. After further investigation I think that `__VA_ARGS__`, and thus the first location, is protected against this assertion firing because:

1. `--c++11` implies `need_variadics()`
2. without `--c++11` (or `--c++20`) defining a macro with ellipsis parameters fails with `ill formed #define directive`
3. without ellipsis no tokens get marked as `EXTPARAMETER`/`OPTPARAMETER`

so I think [the first assertion](https://github.com/jefftrull/wave/blob/60ebbd80e35cd8a7fffc71870f7be22e3a0ad4ce/include/boost/wave/util/cpp_macromap.hpp#L988) is fine. If you can find a test case that shows otherwise I would be glad to see it.

The second one, on the other hand, needs a language check for the case when we run with `--c++11`. I'll put both checks in anyway for the sake of consistency...

Reply:
Thanks for your investigation and that analysis. I think your proposed solution is fine.